### ISSUE TITLE
Scroll Position Logic Improvements

### ISSUE THREAD

**Issue by ryanwiemer** [ryanwiemer/rw#1]
Come up with a better solution for when to remember or not remember the scroll position. 
* https://github.com/gatsbyjs/gatsby/issues/3249
* https://github.com/gatsbyjs/gatsby/pull/3483  

**Remember Scroll Position**
* When a user clicks the back button
* When a user manually refreshes the page

**Start At The Top Of The Page**
* When a user clicks on a link

**Comment by dsbrianwebster** [comment]
How about this Ryan? Seems to satisfy all the above criteria for me when I include in in my gatsby-browser.js file?

```
exports.onRouteUpdate = args => {
  if (typeof window !== 'undefined' && args.action == "PUSH") {
    window.scrollTo(0, 0)
  }
}

exports.shouldUpdateScroll = args => {
  //  scroll on < back, forward >, refresh, but not link clicks
  return (
    args.prevRouterProps == null ||
    (args.prevRouterProps && args.prevRouterProps.history.action == "POP")
  ) ? true : false;
};

```

**Comment by ryanwiemer** [comment]
@dsbrianwebster thanks for the suggestion! After an initial test that appears to accomplish what I was after. I will have to do a little more testing to be sure but the logic looks good to me.

**Comment by ziyafenn** [comment]
Questions:
Do i have to create _gatsby-browser.js_ inside _src_ folder or in the root?

**Comment by ziyafenn** [comment]
Didn't really work for me. I see it working in a console but no effect on scrolling position.

**Comment by ryanwiemer** [comment]
@ziyafenn Hmmm it's hard to say what the issue might be without looking at your code. I have used this solution successfully on two different websites. The best advice I can give is make sure you have the gatsby-browser.js file in the correct location at the root and also be sure to update Gatsby to the latest version. 

If you are still having issues maybe create a post on stackoverflow preferably with a link to your repo.

**Comment by ryanwiemer** [comment]
@ziyafenn, 

Sorry for the double post if you get notified again. Not sure what you issue might be. For good measure I would clear your cache and make sure your are using the latest version of Gatsby. I used Gatsby version 1.9.172 when testing this solution and it seemed to work fine for me. I can't vouch for future or past versions though. 

I thought for a moment I was able to replicate your issue but after checking my dependencies and clearing my cache it was working. 🤷

Sorry if that isn't too helpful.
